feat: add Huawei auth adapter

sudo-happy opened the pull request:
Add huawei auth adapter
I refer to google auth adapter
More information about huawei oauth: [Huawei developer](https://developer.huawei.com/consumer/en/doc/development/HMS-3-References/account-verify-id-token-v3)

### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Related issue: FILL_THIS_OUT

### Approach
<!-- Add a description of the approach in this PR. -->

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [ ] Add tests
- [x] Add changes to documentation (guides, repository pages, in-code descriptions)
- [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [ ] Add new Parse Error codes to Parse JS SDK <!-- no hard-coded error codes in Parse Server -->
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

sudo-happy committed:
Add huawei auth adapter

Add huawei auth adapter
I refer to google auth adapter

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,66 @@
+  
+  // Content Check
+  if( iss !== TOKEN_ISSUER )
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'id token not issued by correct provider')
```
Can you please reformat the error messages into full sentences? We have a PR already that will do that with all other existing error messages.

For example:
> ID token not issued by correct provider.

sudo-happy replied:
I see, I reformat some error messages. ( same case, same message )

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,66 @@
+  const { sub, iss, exp, iat } = huaweiData
+  
+  // Content Check
+  if( iss !== TOKEN_ISSUER )
```
Can you please run lint fix to properly format the code?

sudo-happy replied:
Sorry I'm poor to javascript or babel. I need some info
Now I run command "eslint src/Adapters/Auth/huawei.js --fix"
Is there someting I should do?

mtrezza replied:
that should do it

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,66 @@
+module.exports = {
+  validateAppId: validateAppId,
+  validateAuthData: validateAuthData
+};
```
Please add a new line at the end of the file

sudo-happy committed:
Fix: Error message and etc for PR

1. Run eslint
eslint src/Adapters/Auth/huawei.js --fix

2. Reformat error message
3. Add new line to last of page

mtrezza committed:
Merge branch 'alpha' into alpha

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,65 @@
+    throw new Parse.Error(Parse.Error.OTHER_CAUSE,
+      huaweiData.error_description ?
+        `${huaweiData.error_description} (${huaweiData.error}::${huaweiData.sub_error})` :
+        `undefined Error from huawei (${huaweiData.error}::${huaweiData.sub_error || "No SubError"}})`
```
Could you correct the capitalization to:
> Undefined error from Huawei API ...

and what does "No SubError" mean?

sudo-happy replied:
Oh, Sorry that's my mistake. I will fix it
I thought if there isn't error description.
[doc](https://developer.huawei.com/consumer/en/doc/development/HMSCore-Guides/open-platform-error-0000001053869182) But all errors always have description .

sudo-happy replied:
I remove ambiguous message "No SubError". and simplify code

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,65 @@
+  if(iss !== TOKEN_ISSUER)
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'ID token not issued by correct provider.');
+  if(id !== sub)
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'auth data is invalid for this user.');
```
Could you change capitalization to:
> Auth data is ...

sudo-happy replied:
OK, I will update that
But there is same message "auth data is invalid for this user." in facebook.js and google.js
Can you guide me which right?

mtrezza replied:
We have a PR already that will also correct all other existing error messages. https://github.com/parse-community/parse-server/pull/7446

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,65 @@
+
+  // Expiration Check
+  if(exp < Date.now() / 1000)
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'ID token is expired');
```
Please add dot at the end of this and all other error messages:

> ID token is expired.
> ID token not issued by correct provider or method.

sudo-happy committed:
fix:update huawei error messages

- Add dot to end of message.
- Remove ambiguos message "No SubError"
- Change capitalization

sudo-happy committed:
Merge branch 'parse-community:alpha' into alpha

mtrezza commented on src/Adapters/Auth/huawei.js:
```diff
@@ -0,0 +1,66 @@
+
+async function verifyIdToken({ id_token: token, id }) {
+  if (!token) {
+    throw new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, `id token is invalid for this user.`);
```
Please also capitalize this and all other error messages:
> ID token is invalid for this user.

"ID token" looks confusing, because there seems to be an ID and a token and these do not appear to be the same?

sudo-happy replied:
Sorry, I misunderstood that you said me to change "id token" to "ID token"
Now I change all "ID token" to "id token"

mtrezza replied:
I meant to ask about the term "ID token", which sounds confusing.

I suggest to write the error message as "Token is invalid for this user.", because these are 2 parameters, a `token` and an `id`. The Huawei API key `id_token` seems to just mean "a token for an ID".

sudo-happy replied:
I completely understand
Thank you for your kind comment!!

sudo-happy committed:
fix: remove usless space from eslint

sudo-happy committed:
fix: update error messages

ID token -> id token

sudo-happy committed:
fix: update error message

mtrezza commented:
Do we need any tests here? Do other auth adapters have specific tests?

sudo-happy commented:
> Do we need any tests here? Do other auth adapters have specific tests?

I only use Facebook and Huawei now. These two adapters are working well in common case.
 - success
 - expired token
 - wrong token
 - id is not matched

And my service uses very limited APIs
- Login
- Signup
- Link to other users

I only tested upper cases.
If there isn't features to use auth adapter, I think it works well

mtrezza commented:
I mean whether there are existing automated tests for other auth adapters, that we also need to add for this new Huawei adapter? See the /spec folder.

sudo-happy committed:
Merge branch 'parse-community:alpha' into alpha

mtrezza reviewed (changes requested):
Could you add the tests to cover the adapter? See `TwitterAuth.spec.js` as an example.

mtrezza committed:
Merge branch 'alpha' into alpha
